[Fix #259] Migrate to DB backed sessions

omartell opened the pull request:
Fixes: https://github.com/status-im/open-bounty/issues/259
Status: Finished

## Summary
This migrates from Immutant servlet based sessions to DB backed sessions using [ring.middlware.session](https://ring-clojure.github.io/ring/ring.middleware.session.html), [jdbc-ring-session](https://github.com/luminus-framework/jdbc-ring-session) and [ring-session-timeout](http://ring-clojure.github.io/ring-session-timeout/)

## Why?
- Need to potentially store more data in the session.
- Need to share session information across different application processes.
- A user's session should survive application restarts.

## Things to note:
- A new table `session_store` has been added to hold session information. This will be automatically created when the app starts.
- Any stale session records will be automatically deleted by a session cleaner that runs on the background. A session is considered stale after 6 hours of inactivity.

## QA Notes

(added by @martinklepsch)

- A successful login
- A successful logout
 - After 6 hours of inactivity, they should be logged out and redirected to the home page. The 6 hour period could be extracted to the config and maybe they can change it in their test environment to something shorter (10 min)?
- Session being still present after server restart would be an important thing to test here.


## Alternatives:
I think this might become a bottleneck in the future as every new session hits the DB, but not really sure about this as I don't know the current usage patterns. An alternative could be to use encrypted cookie based sessions, but this depends on the data that's goint to be stored in the session.

I'd appreciate your feedback on the approach as well as on the implementation.

More info:

- http://www.luminusweb.net/docs/sessions_cookies.html
- https://github.com/ring-clojure/ring/wiki/Sessions

martinklepsch commented on project.clj:
```diff
@@ -50,7 +50,9 @@
-                 [cljs-web3 "0.19.0-0-7"]]
+                 [cljs-web3 "0.19.0-0-7"]
+                 [jdbc-ring-session "1.0"]
+                 [ring/ring-session-timeout "0.2.0"]]
```
I think it would be nice to move these closer to the other ring-related dependencies (roughly line 20).

martinklepsch commented on src/clj/commiteth/middleware.clj:
```diff
@@ -64,15 +69,22 @@
+  :start #(start-cleaner *db*)
+  :stop  #(stop-cleaner session-cleaner))
+
+(def redirect-to-home-page (response/redirect "/"))
```
No need for a top-level def here I'd say, just pass `(response/redirect "/")` directly.

martinklepsch commented on src/clj/commiteth/middleware.clj:
```diff
@@ -64,15 +69,22 @@
            (assoc-in [:security :anti-forgery] false)
-           (dissoc :session)))
+           (assoc-in [:session] {:cookie-attrs {:http-only true}
+                                 :store        (jdbc-store *db*)})))
```
Might be missing something here but could we use `ring.middleware.session` for this?

```clj
(ring.middleware.session/wrap-session
   {:cookie-attrs {:http-only true}
    :store        (jdbc-store *db*)})
```

omartell commented:
Thanks for the feedback @martinklepsch! I've addressed all of your comments I think. Let me know if there's something else you'd like me to change.

Regarding the setup, yes it was a bit difficult, particularly getting up and running the ethereum implemention. So, perhaps adding a couple more instructions on how run `geth`, adding an account and dealing with sync issues would help. The other thing I had trouble with was `web3j` as I installed the latest version from homebrew and that didn't work, so I had to manually downgrade.

pablodip commented:
@omartell Thank you for the PR! :)

Could you please describe steps for QA to verify that sessions are working properly? You can do it in the description of the pull request.

omartell commented:
@pablodip in general,  everything should work as normal. 

The things they can double check are probably:
- A successful login
- A successful logout
- After 6 hours of inactivity, they should be logged out and redirected to the home page. The 6 hour period could be extracted to the config and maybe they can change it in their test environment to something shorter (10 min)?

How does that sound?

martinklepsch commented:
- Session being still present after server restart would be an important thing to test here. 

I tested this PR and it works as advertised. @omartell would you mind rebasing against `develop` so we can get this merged?

omartell commented:
@martinklepsch rebased with the latest from `develop`

pablodip commented:
> After 6 hours of inactivity, they should be logged out and redirected to the home page. The 6 hour period could be extracted to the config and maybe they can change it in their test environment to something shorter (10 min)?
> Session being still present after server restart would be an important thing to test here.

These things are't probably clear for QA yet. Could those be simplified or adapted to our current QA tools?

cc @churik

churik commented:
thanks @pablodip 
In order to test this PR I need info where I can change default session time (6 hours is not really convenient for testing)

omartell commented:
@churik you'd need to change it here https://github.com/omartell/open-bounty/blob/feature/db-sessions-%23259/src/clj/commiteth/middleware.clj#L81. The inactivity time needs to be specified in seconds.

siphiuel commented:
This might present a problem for our testing process. I suggest to add a config.edn value for this, like `session-timeout` or something.

omartell commented:
@siphiuel @churik Ok, I've changed the code to get the `session-timeout` value from the `config.edn` file and added the `session-timeout` key to the dev, test, and prod files.

pablodip commented:
@churik Can you change the `config.edn` in the testing environment and restart the app, or isn't that possible yet? Let us know what gaps you find in the QA notes of this issue.

churik commented:
@omartell thanks for contributing!
Can you please take a look into this issues?
## 1. User is not logged out automatically after `session-timeout` is finished (page reload needed)

Steps:
*Requirements*: `session-timeout` is set 
- open https://testing.openbounty.status.im/app (with deployed PR-288)
- log in
- switch to another tab
- after `session-timeout` + 10 sec switch to https://testing.openbounty.status.im/app again

### Actual result:
logged out if you reload page manually

### Expected result: 
automatic log out

**Video:** http://take.ms/I1Nnd
**OS:**  Mac OS High Sierra
**Browser:** Chrome 64

## 2. JS errors in console
Steps:
- open https://testing.openbounty.status.im/app (with deployed PR-288)
- open JS console, wait for 2-3 min

### Actual result:
*Reproducible when user is logged out *
![2js](https://user-images.githubusercontent.com/4557972/37509339-a2ffb568-28ff-11e8-8756-0eb352061270.png)
![1js](https://user-images.githubusercontent.com/4557972/37509340-a33780ec-28ff-11e8-8f60-96feedb328f9.png)

### Expected result: 
no errors;
 in fresh `develop` there are no such JS errors.


**OS:**  Mac OS High Sierra
**Browser:** Chrome 64

martinklepsch commented:
> 1. User is not logged out automatically after session-timeout is finished (page reload needed)

Did this work with the in-memory session store? If not I feel like this should be tackled as a separate bounty since it hasn't been explicitly stated in the acceptance criteria? 

> 2. JS errors in console

Can you reproduce this consistently? I tried on https://testing.openbounty.status.im/app (but not 100% sure it was still on PR-288) and couldn't trigger this issue once. 

Also this PR didn't change any ClojureScript code so I'm not sure if this error really comes from it?

churik commented:
@martinklepsch 
2. no, other PR is deployed on `testing env`. 
With last `develop` branch on `staging env` is not reporoducible

martinklepsch commented:
@churik could you test this one again? It's been lingering for a long time and my feeling is that the JS error was due to something else... Thanks! 🙌

churik commented:
It is not working for me at all.

Steps: 
- deploy PR-288 to `testing env`
- edit `config.edn` in `testing.openbounty.status.im` vm (set `:session-timeout`  to  180}
- restart docker container
- log in to https://testing.openbounty.status.im, wait for 200 sec.
- refresh  https://testing.openbounty.status.im

I wasn't logged out.
Did I miss something? @martinklepsch 
It worked when I tested this PR for the first time.

martinklepsch commented:
Moved this PR to #288 so we can rebase against develop if needed. Please direct further testing there @churik 🙂
